### PR TITLE
Variant Creator

### PR DESCRIPTION
**Pull request by breakone**
| Q             | A
| ------------- | ---
| Bug fix?      |no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #...   

Create variants of a product based on attributes in the backend UI.

### CODE REVIEW

File: src/CoreShop/Bundle/CoreBundle/Resources/public/pimcore/js/resource.js [hunk 1]
```diff
@@ -189,6 +189,143 @@ coreshop.core.resource = Class.create(coreshop.resource, {
...
+
+                                    if (!hasCheckedLeaf) {
+                                        allParentsHaveSelection = false;
+                                        return false;
+                                    }
+                                });
+
+                                if (allParentsHaveSelection) {
+                                    applyButton.enable();
+                                    //Ext.Msg.alert('Success', 'At least one leaf from each parent node is selected.');
```
Comment (dpfaffenbauer) [inline src/CoreShop/Bundle/CoreBundle/Resources/public/pimcore/js/resource.js]: Alert not needed?

File: src/CoreShop/Bundle/CoreBundle/Resources/public/pimcore/js/resource.js [hunk 2]
```diff
@@ -189,6 +189,143 @@ coreshop.core.resource = Class.create(coreshop.resource, {
...
+                                } else {
+                                    applyButton.disable();
+                                    //Ext.Msg.alert('Error', 'Please select at least one leaf from each parent node.');
```
Comment (dpfaffenbauer) [inline src/CoreShop/Bundle/CoreBundle/Resources/public/pimcore/js/resource.js]: Alert not needed?

File: src/CoreShop/Bundle/VariantBundle/Messenger/Handler/CreateVariantMessageHandler.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        $attributeIds = $message->getAttributeIds();
+        if (!$attributeIds) {
+            return;
+        }
+
+        $variant = $this->variantGeneratorService->generateVariant($attributeIds, $object);
+
+        // TODO: check if needed, because of too many notifications for massive variant generation
```
Comment (dpfaffenbauer) [inline src/CoreShop/Bundle/VariantBundle/Messenger/Handler/CreateVariantMessageHandler.php]: todo is still a todo?

File: src/CoreShop/Bundle/VariantBundle/Service/VariantGeneratorService.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        return $variants;
+    }
+
+    public function generateVariant(array $attributeIds, ProductVariantAwareInterface $product): ?ProductVariantAwareInterface
+    {
+        if (count($product->getAllowedAttributeGroups()) !== count($attributeIds)) {
+            return null;
+        }
+
+        $class = get_class($product);
```
Comment (dpfaffenbauer) [inline src/CoreShop/Bundle/VariantBundle/Service/VariantGeneratorService.php]: should we check for Pimcore Concrete here and use `$product::getListing()` here?

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

File: src/CoreShop/Bundle/VariantBundle/Service/VariantGeneratorService.php (hunk 4 above)
Comment (breakone) [reply src/CoreShop/Bundle/VariantBundle/Service/VariantGeneratorService.php]: @dpfaffenbauer absolutely

**Review by dpfaffenbauer — APPROVED** [review]

**Pull request merged**